src/coreclr/gc/gc.cpp

```diff
@@ -13696,12 +13712,21 @@ void gc_heap::distribute_free_regions()
     }
 
 #ifdef MULTIPLE_HEAPS
-    for (int kind = basic_free_region; kind < count_free_region_kinds; kind++)
+    if (joined_last_gc_before_oom) // high_memory_load_p)
     {
-        if (global_regions_to_decommit[kind].get_num_free_regions() != 0)
+        while (decommit_step(DECOMMIT_TIME_STEP_MILLISECONDS))
```

Comment:
decommit a percentage?

or compare against the hard limit, etc. and make "good" progress